File: 0046-permutations/0046-permutations.java
import java.util.*;
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(new ArrayList<>(), nums, nums.length,  ans);
        return ans;
    }
    public static void backtrack(List<Integer> curr, int[] nums, int k, List<List<Integer>> ans) {
        if(curr.size() == k) {
            ans.add(new ArrayList<>(curr));
            return ;
        }
        for(int i= 0; i < nums.length; i++) {
            if(!curr.contains(nums[i]))  {
                curr.add(nums[i]);
                backtrack(curr, nums,k, ans);
                curr.remove(curr.size()-1);
            }
        }
    }
}

File: 0046-permutations/0046-permutations.cpp


#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr;
        backtrack(curr, nums, ans);
        return ans;
    }

private:
    void backtrack(vector<int>& curr, vector<int>& nums, vector<vector<int>>& ans) {
        int n = nums.size();
        // Base case
        if (curr.size() == n) {
            ans.push_back(curr); // Add a copy of current permutation to the answer
            return;
        }
        for (int num : nums) { // Iterate over nums elements
            if (find(curr.begin(), curr.end(), num) == curr.end()) { // If num is not in curr
                curr.push_back(num);
                backtrack(curr, nums, ans);
                curr.pop_back(); // Backtrack
            }
        }
    }
};

